lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,546 @@
+        """
+        if "path" in kwargs:
+            path = Path(kwargs.get("path"))
+            if path.is_symlink():
+                path = Path(os.path.realpath(path))
+            if "name" not in kwargs:
+                name = path.name
+                if name.endswith(".tar.xz"):
+                    name = name[:-7]
+                kwargs["name"] = name
+            if "controller" not in kwargs:
+                kwargs["controller"] = path.parent.name
+            del kwargs["path"]  # Don't let the parental unit see this
+
+    @staticmethod
+    def create(**kwargs):
+        """
+        create A simple factory method to construct a new Dataset object and
+        add it to the database. Errors are hidden here, and we return None.
+
+        Args:
+            kwargs (dict):
+                "user": The owner of the dataset; defaults to "public".
+                "path": A tarball file path from which the controller (host)
+                    name, the tarball dataset name (basename minus extension)
+                    or both will be derived. The `path` key will be removed
+                    from the dict after processing. (The SQLAlchemy super
+                    constructor would complain about it as its not a column.)
+                "controller": The controller name (hostname) of the dataset;
+                    this is retained if specified, or will be constructed
+                    from "path" if not present.
+                "name": The dataset name (file path basename minus ".tar.xz");
+                    this is retained if specified, or will be constructed from
+                    "path" if not present.
+                "state": The initial state of the new dataset.
+
+        Returns:
+            Dataset: A new Dataset object initialized with the keyword
+            parameters. (See `__init__` for details.)
+        """
+        try:
+            dataset = Dataset(**kwargs)
+            dataset.add()
+        except Exception as e:
+            Dataset.logger.error("Failed create: {}", str(e))
+            return None
+        else:
+            return dataset
+
+    @staticmethod
+    def attach(**kwargs):
+        """
+        attach Attempt to fetch dataset for the controller and dataset name,
+        or using a specified file path (see __init__ and _render_path).
+
+        If the `state` kwarg is specified, the dataset will be created in that
+        state or, if it already exists, `attach` will attempt to advance the
+        dataset to that state.
+
+        Args:
+            kwargs (dict):
+                "path": A tarball file path from which the controller (host)
+                    name, the tarball dataset name (basename minus extension)
+                    or both will be derived. The `path` key will be removed
+                    from the dict after processing. (The SQLAlchemy super
+                    constructor would complain about it as its not a column.)
```

Comment:
Ping?

Reply:
Speaking generally, I think your original approach of warning every caller that the dictionary object was going to be modified was entirely appropriate!  In Python, argument passing and assignment in general is, um, somewhere between weird and counterintuitive, and so a caller handing a dictionary to a function won't necessarily expect the dictionary to come back modified (unless that's the express purpose of the function, of course).  And, so I think it was good to post the warning.

However, following the principal of least surprise, it would be better to avoid the need for the warning (and then you won't have to issue it, and it won't trip up any callers who actually care, unlike the ones in the present case).

And, copying the dict shouldn't be a big deal, because it's only got a few entries in it.  (It probably costs less than setting up the call to the SQLAlchemy function.)